### DISCUSSION TITLE
Artillery version 2.0.0-37 publish to Open Telemetry - Questions

### DISCUSSION THREAD

**Discussion by Thilaknath** [artilleryio/artillery#2128] [Artillery]
Hello Team,

Congratulations with the new release! 
I have installed the above version and tried to send the performance test data to our open telelmetry instance which is hosted in our organization since I am having issues with dynatrace.

I have the following configuration
```
    - type: "open-telemetry"
      serviceName: "async-service"
      metrics:
        exporter: "otlp-http"
        endpoint: "ingest-metrics-otlp.cls-03.cloud.logs.services.com:443"
        headers:
          Authorization: "Basic XXX-XXX-XXX"
        attributes:
          environment: "performance"
      traces:
        exporter: "otlp-proto"
        endpoint: "ingest-traces-otlp.cls-03.cloud.logs.services.com:443"
        headers:
          Authorization: "Basic XXX-XXX-XXX"
        useRequestNames: true
        attributes:
          environment: "performance"
          tool: "Artillery"
```

The Authorization header provided is in this format 
The [INGEST_AUTH_HEADER] is the Base64 encoded version of username and password. Specifically, it is composed this way:
"Basic " + base64([INGEST_USERNAME] + ":" + [INGEST_PASSWORD])

I have also turned on the debug flag as suggested here https://www.artillery.io/docs/reference/extensions/publish-metrics#opentelemetry

`DEBUG=plugin:publish-metrics:open-telemetry artillery run my-script.yaml`

My script executes fine, But I don't see any data being sent to the open telemetry collector and also no information is shown in the debug logs on what happens during the publish metrics step. Can you have a look?

**Comment by Thilaknath** [comment]
Update: I executed a test providing no credentials, the plugin does not throw any error and doesn't output any information to the logs. I have the `DEBUG` set. Any pointers to this would be great

**Reply by InesNi** [reply]
Hi @Thilaknath,

I'm working on making debug better, I'll to try to have a version out soon with better debugging capabilities, as OTeL doesn't expose a lot of it by default.

- So to confirm, you're sending this to an OpenTelemetry collector instance from your org, right?
- Can you try adding the protocol (http/https) in the endpoint?

**Reply by Thilaknath** [reply]
Yes @InesNi I am sending to the otel collector at my org. Do we need to specify any index on where the metrics should be shipped to in the above config? 
By default we already have a index pattern metrics-* and metrics-otel* in our open search instance.
I will try adding the protocol and give it a try.

**Reply by Thilaknath** [reply]
Update: Trying both http and https, I did not seem to receive the metrics at my end.

**Reply by InesNi** [reply]
Hi @Thilaknath.

What about `traces` with the `otlp-proto` exporter from the config above? Are you able to see those? If so, can you also try ingesting metrics through `otlp-proto`?

**Reply by Thilaknath** [reply]
@InesNi Have tried with the `otlp-proto` with the combination of having `endpoint` with protocols (`http or https`) and without them. Not seeing the data ingested to the metrics or traces.

**Reply by Thilaknath** [reply]
@InesNi  is it possible to use the username/password directly ?

**Reply by InesNi** [reply]
Hi @Thilaknath,

The improved debug for OpenTelemetry reporter is available now on the canary version `artillery@2.0.0-37-889006a`. Hopefully it will help you get more information on what might be going on.

Artillery uses OpenTelemetry OTLP [http](https://www.npmjs.com/package/@opentelemetry/exporter-metrics-otlp-http) and [proto](https://www.npmjs.com/package/@opentelemetry/exporter-metrics-otlp-proto) exporters for metrics, and equivalent for traces, and only exposes the optional `endpoint` and `headers` setting.

You would have to configure authentication within your collector if you are using OTel Collector.
I hope the debug output helps to troubleshoot better, and would also look into OpenTelemetry Collector docs to see if there is something that needs to be configured additionally. Looking into [`otlp` reciever](https://github.com/open-telemetry/opentelemetry-collector/blob/main/receiver/otlpreceiver/README.md) docs and [HTTP configuration settings](https://github.com/open-telemetry/opentelemetry-collector/blob/main/config/confighttp/README.md) might be a good place to check for clues.

**Reply by Thilaknath** [reply]
Hello @InesNi Thank you . The debugging definitely helps. Something I noticed is that our  ingestUrl from open telemetry service looks something like this below
`ingest-metrics-otlp-sf-XXXX-XXXX-XXXX-XXXX.cls-03.cloud.logs.services.ondemand.com:443` . 

This is the URL I am advised to write metrics data to. I have a similar one for `traces` too.

Now with the above URL specified as is with the protocol `otlp-http` for metrics and `otlp-proto` for traces. For some reason, it tries to connect through the local host. Refer logs below

```
{"stack":"Error: PeriodicExportingMetricReader: metrics export failed (error Error: connect ECONNREFUSED 127.0.0.1:443)\n    at doExport (/usr/local/lib/node_modules/artillery/node_modules/@opentelemetry/sdk-metrics/build/src/export/PeriodicExportingMetricReader.js:76:23)\n    at processTicksAndRejections (node:internal/process/task_queues:96:5)\n    at async PeriodicExportingMetricReader._doRun (/usr/local/lib/node_modules/artillery/node_modules/@opentelemetry/sdk-metrics/build/src/export/PeriodicExportingMetricReader.js:84:13)\n    at async PeriodicExportingMetricReader._runOnce (/usr/local/lib/node_modules/artillery/node_modules/@opentelemetry/sdk-metrics/build/src/export/PeriodicExportingMetricReader.js:55:13)\n    at async EventEmitter.<anonymous> (/usr/local/lib/node_modules/artillery/node_modules/artillery-plugin-publish-metrics/lib/open-telemetry/index.js:104:9)","message":"PeriodicExportingMetricReader: metrics export failed (error Error: connect ECONNREFUSED 127.0.0.1:443)","name":"Error"}
```

If I try to remove the port from the `endpoint` and specify `http` as the protocol -> I get 
```
{"stack":"OTLPExporterError: Request Timeout\n    at ClientRequest.<anonymous> (/usr/local/lib/node_modules/artillery/node_modules/@opentelemetry/otlp-exporter-base/build/src/platform/node/util.js:114:29)\n    at ClientRequest.emit (node:events:527:28)\n    at Socket.socketCloseListener (node:_http_client:427:11)\n    at Socket.emit (node:events:539:35)\n    at TCP.<anonymous> (node:net:709:12)","message":"Request Timeout","name":"OTLPExporterError","code":"ECONNRESET"}
```

If I specify `https` as the protocol -> I get `NOT FOUND` 

Also something I noticed. While the metrics is being prepped. I get the following errors
```
Invalid metric name: "plugins.metrics-by-endpoint./actuator/health.codes.200". The metric name should be a ASCII string with a length no greater than 255 characters.
Invalid metric name: "plugins.metrics-by-endpoint.response_time./actuator/health.min". The metric name should be a ASCII string with a length no greater than 255 characters.
Invalid metric name: "plugins.metrics-by-endpoint.response_time./actuator/health.max". The metric name should be a ASCII string with a length no greater than 255 characters
```

**Reply by InesNi** [reply]
Hi @Thilaknath,

Thank you for the detailed information 🙏🏻 - the logs are proving useful indeed! I probably won't be able to look at this tomorrow, but I intend to circle back to it on Friday.

In the meantime, could you please try to run with `https` **and** the `port`, i.e.:
`https://ingest-metrics-otlp-sf-XXXX-XXXX-XXXX-XXXX.cls-03.cloud.logs.services.ondemand.com:443`
as I don't think you mentioned that variation above. And if possible let me know what the logs say for that one if it doesn't work again.

---

Regarding the others errors you mentioned, it should be the same issue as Dynatrace, so the same [workaround I mentioned there](https://github.com/artilleryio/artillery/discussions/2064#discussioncomment-6995749) should work here. Opentelemetry are actually in the process of allowing `/` characters in their JS implementation, so this problem should go away soon 🤞🏻. 

Thanks again for your patience!

**Reply by Thilaknath** [reply]
Hello @InesNi 
Thanks for your support. Here are the logs from all the scenario.

In my config for the below results.  Metrics has the exporter set to `otlp-http` and traces have the exporter set to `otlp-proto`. Errors observed below

**With https and 443**
```
{"stack":"Error: PeriodicExportingMetricReader: metrics export failed (error OTLPExporterError: Not Found)
   at doExport (/usr/local/lib/node_modules/artillery/node_modules/@opentelemetry/sdk-metrics/build/src/export/PeriodicExportingMetricReader.js:76:23)
   at processTicksAndRejections (node:internal/process/task_queues:96:5)
   at async PeriodicExportingMetricReader._doRun (/usr/local/lib/node_modules/artillery/node_modules/@opentelemetry/sdk-metrics/build/src/export/PeriodicExportingMetricReader.js:84:13)
   at async PeriodicExportingMetricReader._runOnce (/usr/local/lib/node_modules/artillery/node_modules/@opentelemetry/sdk-metrics/build/src/export/PeriodicExportingMetricReader.js:55:13)
   at async EventEmitter.<anonymous> (/usr/local/lib/node_modules/artillery/node_modules/artillery-plugin-publish-metrics/lib/open-telemetry/index.js:104:9)",
   "message":"PeriodicExportingMetricReader: metrics export failed (error OTLPExporterError: Not Found)","name":"Error"}
```

 **With http and 443 (Un wanted scenario)**
```
{"stack":"OTLPExporterError: Bad Request
   at IncomingMessage.<anonymous> (/usr/local/lib/node_modules/artillery/node_modules/@opentelemetry/otlp-exporter-base/build/src/platform/node/util.js:103:39)
   at IncomingMessage.emit (node:events:539:35)
   at endReadableNT (node:internal/streams/readable:1345:12)
   at processTicksAndRejections (node:internal/process/task_queues:83:21)","message":"Bad Request","name":"OTLPExporterError",
   "data":"<html>\r\n<head><title>400 The plain HTTP request was sent to HTTPS port</title></head>\r\n<body>\r\n<center><h1>400 Bad Request</h1></center>\r\n<center>The plain HTTP request was sent to HTTPS port</center>\r\n<hr><center>nginx</center>\r\n</body>\r\n</html>\r\n","code":"400"}
```

**With just http and no port**
```
⠏ {"stack":"OTLPExporterError: Request Timeout
   at ClientRequest.<anonymous> (/usr/local/lib/node_modules/artillery/node_modules/@opentelemetry/otlp-exporter-base/build/src/platform/node/util.js:114:29)
   at ClientRequest.emit (node:events:527:28)
   at Socket.socketCloseListener (node:_http_client:427:11)
   at Socket.emit (node:events:539:35)
   at TCP.<anonymous> (node:net:709:12)","message":"Request Timeout","name":"OTLPExporterError","code":"ECONNRESET"}
```

**With Just the URL with no protocol and port**
```
{"stack":"Error: PeriodicExportingMetricReader: metrics export failed (error TypeError [ERR_INVALID_URL]: Invalid URL)...
```

**With Just https and no port**
```
  {"stack":"OTLPExporterError: Not Found
   at IncomingMessage.<anonymous> (/usr/local/lib/node_modules/artillery/node_modules/@opentelemetry/otlp-exporter-base/build/src/platform/node/util.js:103:39)
   at IncomingMessage.emit (node:events:539:35)
   at endReadableNT (node:internal/streams/readable:1345:12)
   at processTicksAndRejections (node:internal/process/task_queues:83:21)","message":"Not Found","name":"OTLPExporterError","data":"Status: 404\nDescription: Not Found\n","code":"404"}
```

**Reply by Thilaknath** [reply]
Update:
This seems to be the ideal config that was suggested from the team that is managing our open search instance

```
 - type: "open-telemetry"
      serviceName: "async-service"
      metrics:
        exporter: "otlp-grpc"
        endpoint: "https://ingest-metrics-otlp.cls-03.cloud.logs.services.com:443"
        headers:
          Authorization: "Basic XXX-XXX-XXX"
        attributes:
          environment: "performance"
      traces:
        exporter: "otlp-proto"
        endpoint: "https://ingest-traces-otlp.cls-03.cloud.logs.services.com:443"
        headers:
          Authorization: "Basic XXX-XXX-XXX"
        useRequestNames: true
        attributes:
          environment: "performance"
          tool: "Artillery"
```

With the above, I am able to see a new index created in our open search dashboard. But it does not have any metrics. Error observed

```
{"stack":"Error: PeriodicExportingMetricReader: metrics export failed (error Error: 13 INTERNAL: Received RST_STREAM with code 0)
    at doExport (/usr/local/lib/node_modules/artillery/node_modules/@opentelemetry/sdk-metrics/build/src/export/PeriodicExportingMetricReader.js:76:23)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)    at async PeriodicExportingMetricReader._doRun (/usr/local/lib/node_modules/artillery/node_modules/@opentelemetry/sdk-metrics/build/src/export/PeriodicExportingMetricReader.js:84:13)    
    at async PeriodicExportingMetricReader._runOnce (/usr/local/lib/node_modules/artillery/node_modules/@opentelemetry/sdk-metrics/build/src/export/PeriodicExportingMetricReader.js:55:13)    
    at async EventEmitter.<anonymous> (/usr/local/lib/node_modules/artillery/node_modules/artillery-plugin-publish-metrics/lib/open-telemetry/index.js:104:9)",
    "message":"PeriodicExportingMetricReader: metrics export failed (error Error: 13 INTERNAL: Received RST_STREAM with code 0)","name":"Error"}
```

**Reply by InesNi** [reply]
Hi @Thilaknath,

Thanks for the update.

I believe now that you're sending metrics via `otlp-grpc` exporter, you shouldn't be setting headers directly in the test script, but can only set them through environment variables. I've tried to send to a cloud provider (honeycomb) using `otlp-grpc` and setting `headers` and also ran into an issue where it mentions `Headers cannot be set when using grpc` in the debug. Perhaps you also ran into that but missed it in the debug? 

Can you try doing the following:
- Either with `https://ingest-metrics-otlp.cls-03.cloud.logs.services.com:443` or `ingest-metrics-otlp.cls-03.cloud.logs.services.com:443` (I used without `https`), can you try to set your Authorization header using the environment variable `OTEL_EXPORTER_OTLP_METRICS_HEADERS`? For example: `OTEL_EXPORTER_OTLP_METRICS_HEADERS="Authorization=Basic XXXX-XXX"`. (note: the headers must follow [this specification](https://github.com/w3c/baggage/blob/main/baggage/HTTP_HEADER_FORMAT.md))

To set it you can just run the following command in the console:
```shell
export OTEL_EXPORTER_OTLP_METRICS_HEADERS="Authorization=Basic XXXX-XXX"
```
- In the debug information you mentioned above, you mentioned `metrics`, but you didn't mention if you were able to get `traces`. Have you been able to see any traces reaching your platform? 

Thank you!

**Reply by Thilaknath** [reply]
Hello @InesNi Thank you for your response. I do see some metrics now being sent to my open search instance after doing the following 
The protocol remains `otlp-grpc`. The url is `https://ingest-metrics-otlp.cls-03.cloud.logs.services.com:443 or ingest-metrics-otlp.cls-03.cloud.logs.services.com:443` and the Authorization set from the console as advised above. 

The traces are still not visible and I see the following error. I believe we are almost there in getting the correct configuration.
For the traces, 
Without `https`
```
{"stack":"Error: connect ECONNREFUSED 127.0.0.1:443\n    at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1187:16)","message":"connect ECONNREFUSED 127.0.0.1:443","errno":"-61","code":"ECONNREFUSED","syscall":"connect","address":"127.0.0.1","port":"443","name":"Error"}
```

With `https` the following error
```
⠏{"stack":"OTLPExporterError: Not Found\n    at IncomingMessage.<anonymous> (/usr/local/lib/node_modules/artillery/node_modules/@opentelemetry/otlp-exporter-base/build/src/platform/node/util.js:103:39)\n    at IncomingMessage.emit (node:events:539:35)\n    at endReadableNT (node:internal/streams/readable:1345:12)\n    at processTicksAndRejections (node:internal/process/task_queues:83:21)","message":"Not Found","name":"OTLPExporterError","data":"Status: 404\nDescription: Not Found\n","code":"404"}
```

**Reply by Thilaknath** [reply]
@InesNi  Unfortunately my service provider is switching to mTLS very soon. Is it possible to use client certificates to authenticate ?

**Reply by InesNi** [reply]
Hi @Thilaknath,

To answer your last question, at the moment it is not possible to pass the client certificates through Artillery's test script, but the exporters check for them in the [environment variables](https://opentelemetry.io/docs/specs/otel/protocol/exporter/#configuration-options) so I believe you should be able to set them through those.

For mTLS you will need to set the following environment variables:
- To set the same credentials for both traces and metrics:
  - `OTEL_EXPORTER_OTLP_CERTIFICATE`
  - `OTEL_EXPORTER_OTLP_CLIENT_KEY`
  - `OTEL_EXPORTER_OTLP_CLIENT_CERTIFICATE`

- To set signal specific credentials insert `TRACES` or `METRICS` respectively e.g.:
  - `OTEL_EXPORTER_OTLP_TRACES_CLIENT_KEY` for traces
  -`OTEL_EXPORTER_OTLP_METRICS_CLIENT_KEY` for metrics

[Here](https://www.npmjs.com/package/@opentelemetry/exporter-trace-otlp-grpc?activeTab=readme) at the bottom you will find detailed description of each variable - note that the signal specific variables are only specified for traces here but the same description applies to metric equivalent.


---

As for the traces, any progress there? On this, I'm not really sure, it seems to be a problem more on the collector side - perhaps you can try to talk to the team on your side and see if they see anything on the collector logs?

If not, then the only thing I can suggest is try a few things:
- Use `otlp-grpc` exporter instead if supported on your side. We now support `grpc` on our side for `traces` too - this version should have it: `artillery@2.0.0-37-b8419d2`;
- Since you're getting a 404, maybe your endpoint is somehow wrong. Maybe try appending `/v1/traces` to the end of the endpoint? Long shot but nothing to lose. 🤷🏻‍♀️
- Maybe try using the environment variables for the headers (`OTEL_EXPORTER_OTLP_TRACES_HEADERS`) instead of setting them directly in the config. Although this one shouldn't make a difference, it's worth trying if nothing else works.

**Reply by Thilaknath** [reply]
@InesNi Sure noted, Thanks. I did not try to traces. But will give the above suggestions a shot
